rs/sns/cli/src/upgrade_sns_controlled_canister.rs

```diff
@@ -0,0 +1,609 @@
+        std::io::stdout().flush().unwrap();
+        let candid_arg_bytes = encode_upgrade_args(candid_service, args).unwrap();
+        println!("✔️");
+
+        candid_arg_bytes
+    } else {
+        eprintln!(
+            "\n\
+            ⚠️ Skipping upgrade argument validation: Wasm file has no Candid definition! \n\
+            ⚠️ Please consider adding it as follows:\n\
+            cargo install ic-wasm\n\
+            ic-wasm -o augmented-{} {} metadata -v public candid:service -f service.did",
+            wasm.path(),
+            wasm.path(),
+        );
+
+        // Proceed with whatever argument the user has specified without validation.
+        args.map(|args| encode_upgrade_args_without_service(args).unwrap())
+    };
+
+    std::io::stdout().flush().unwrap();
+    std::io::stderr().flush().unwrap();
+
+    Ok(canister_arg)
+}
+
+/// Checks if `canister_id` is an ID of a canister that exists and has some Wasm code installed.
+///
+/// In the Ok result, returns a tuple with the following components:
+/// 1. Set of controllers.
+/// 2. Module hash of the installed code.
+///
+/// See https://internetcomputer.org/docs/current/references/ic-interface-spec#state-tree-canister-information
+///
+/// This function is analogous to `dfx canister info`.
+pub async fn fetch_canister_info(
+    agent: &Agent,
+    canister_id: CanisterId,
+) -> Result<(BTreeSet<PrincipalId>, Vec<u8>)> {
+    let module_hash = agent
+        .read_state_canister_info(canister_id.get().0, "module_hash")
+        .await
+        .context("Cannot read target canister's module hash.")?;
+
+    let controllers_blob = agent
+        .read_state_canister_info(canister_id.get().0, "controllers")
+        .await
+        .context("Cannot read canister controllers.")?;
+
+    let cbor: Value = serde_cbor::from_slice(&controllers_blob)
+        .expect("Invalid cbor data for controller controllers.");
+
+    let Value::Array(controllers) = cbor else {
+        bail!("Expected controllers to be an array, but got {cbor:?}");
+    };
+
+    let (controllers, errors): (BTreeSet<_>, Vec<_>) =
+        controllers.into_iter().partition_map(|value| {
+            let Value::Bytes(bytes) = value else {
+                let err =
+                    format!("Expected canister controller to be of type bytes, got {value:?}",);
+                return Either::Right(err);
+            };
+            match Principal::try_from(&bytes) {
+                Err(err) => {
+                    let err = format!("Cannot interpret canister controller principal: {err}");
+                    Either::Right(err)
+                }
+                Ok(principal) => Either::Left(PrincipalId(principal)),
+            }
+        });
+
+    if !errors.is_empty() {
+        let err = format!(
+            "Problems with canister controllers:\n  - {}",
+            errors.join("\n  - ")
+        );
+        bail!(err);
+    }
+
+    Ok((controllers, module_hash))
+}
```

Comment:
It's a bit of a shame to duplicate this code between the IC repo and dfx. (Also nice, your code is better than the code in dfx IMO.) I wonder if we could contribute this to ic-agent since it seems useful to have generally (we also do this when proposing an SNS, so this is clearly useful, not sure why there's no easy way to do it).